rpcs3/Emu/Cell/Modules/cellMic.cpp

```diff
@@ -134,45 +133,77 @@ void mic_context::operator()()
 	// Cleanup
 	for (auto& mic_entry : mic_list)
 	{
-		mic_entry.second.close_microphone();
+		mic_entry.close_microphone();
 	}
 }
 
 void mic_context::load_config_and_init()
 {
-	auto device_list = fmt::split(g_cfg.audio.microphone_devices.to_string(), {"@@@"});
+	const std::vector<std::string> device_list = fmt::split(g_cfg.audio.microphone_devices.to_string(), {"@@@"});
 
 	if (!device_list.empty() && mic_list.empty())
 	{
-		switch (g_cfg.audio.microphone_type)
+		// We only register the first device. The rest is registered with cellAudioInRegisterDevice.
+		if (g_cfg.audio.microphone_type == microphone_handler::singstar)
 		{
-		case microphone_handler::standard:
-		{
-			for (s32 index = 0; index < static_cast<s32>(device_list.size()); index++)
-			{
-				mic_list.emplace(std::piecewise_construct, std::forward_as_tuple(index), std::forward_as_tuple(microphone_handler::standard));
-				::at32(mic_list, index).add_device(device_list[index]);
-			}
-			break;
-		}
-		case microphone_handler::singstar:
-		{
-			mic_list.emplace(std::piecewise_construct, std::forward_as_tuple(0), std::forward_as_tuple(microphone_handler::singstar));
+			mic_list.emplace_back(microphone_handler::singstar);
 			::at32(mic_list, 0).add_device(device_list[0]);
+
+			// Singstar uses the same device for 2 players
 			if (device_list.size() >= 2)
+			{
 				::at32(mic_list, 0).add_device(device_list[1]);
-			break;
+			}
 		}
-		case microphone_handler::real_singstar:
-		case microphone_handler::rocksmith:
+		else
 		{
-			mic_list.emplace(std::piecewise_construct, std::forward_as_tuple(0), std::forward_as_tuple(g_cfg.audio.microphone_type));
-			::at32(mic_list, 0).add_device(device_list[0]);
-			break;
+			[[maybe_unused]] const u32 index = register_device(device_list[0]);
 		}
-		case microphone_handler::null:
-		default: break;
+	}
+}
+
+u32 mic_context::register_device(const std::string& device_name, const std::string& name_2_for_singstar)
+{
+	const usz index = mic_list.size();
+
+	switch (g_cfg.audio.microphone_type)
+	{
+	case microphone_handler::standard:
+	case microphone_handler::real_singstar:
+	case microphone_handler::rocksmith:
+	{
+		mic_list.emplace_back(g_cfg.audio.microphone_type.get());
+		::at32(mic_list, index).add_device(device_name);
+
+		if (auto mic_queue = lv2_event_queue::find(event_queue_key))
+		{
+			mic_queue->send(event_queue_source, CELLMIC_ATTACH, index, 0);
 		}
+
+		break;
+	}
+	case microphone_handler::singstar:
+	case microphone_handler::null:
+	default:
+		break;
+	}
+
+	return ::narrow<u32>(index);
+}
+
+void mic_context::unregister_device(u32 dev_num)
```

Comment:
Hmm so let's say a game registers 2 extra mics.
We'd have mic 0, 1, 2. Then if it unregisters the mic with device 1 the device 2 mic becomes the device 1 mic?
I think that's probably incorrect.

Reply:
Changed to use a fixed amount of mics.
We'll see if this works...